File: TrieImplement.java
public class TrieImplement {



        // TrieNode class
        class TrieNode {

            TrieNode[] children;

            boolean isEnd;

            // TrieNode constructor
            public TrieNode() {

                this.children = new TrieNode[26];
            }
        }

        private TrieNode root;

        //Trie Constructor
        public TrieImplement() {

            //a null root trie node
            this.root = new TrieNode();
        }

        //INSERT
        public void insert(String word) {

            TrieNode curr = root;

            for(int i = 0; i < word.length(); i++) { //O(L)

                char c = word.charAt(i);

                // if any appropriate word character is missing in TrieNode
                if(curr.children[c - 'a'] == null) {

                    // crreate a new TrieNode for insertion
                    curr.children[c - 'a'] = new TrieNode();
                }
                //move current from root to children
                curr = curr.children[c - 'a'];
            }
            //set boolean at current TrieNode end to zero
            curr.isEnd = true;
        }

        //SEARCH
        public boolean search(String word) { //O(L)

            TrieNode curr = root;

            for(int i = 0; i < word.length(); i++) {

                char c = word.charAt(i);

                // if any appropriate word character is missing in TrieNode, word not found
                if(curr.children[c - 'a'] == null) return false;

                // update current TrieNode
                curr = curr.children[c - 'a'];
            }
            // at times full length of word is present, but it might not have inserted priorly (isEnd != true), then return only false. Therefore, boolean of current TrieNode
            return curr.isEnd;
        }

        public boolean startsWith(String prefix) { //O(L)

            TrieNode curr = root;

            for(int i = 0; i < prefix.length(); i++) {

                char c = prefix.charAt(i);

                if(curr.children[c - 'a'] == null) return false;

                curr = curr.children[c - 'a'];
            }
            //once loop is finished, given prefix is covered
            return true;
        }


}

/*
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

/*
TIME COMPLEXITY = O(L)

L = length of word
N = number of words

SPACE COMPLEXITY = O(N*L)

For user oriented functions, space complexity = O(1)
*/

File: trie_implement.py
"""
Description: Implement a Trie

Time Complexicity: O(n)
Space Complexicity: O(n)
where n is the length of a given word (string)

Approach:
1. TrieNode
- Create a new treenode with default list of 26 items (all None to start, number of lower letters), and isEnd
- isEnd is the stopping point for any given word (subword)
2. Use TrieNode
+ for insert, start with current node as root
  - traverse the current with each character, if character is not found at a given index -> add a TrieNode
  - in each step update current to the created or available TrieNode for the current character
+ for search and startswith a prefix
  - traverse the current with each character, if character is not found at a given index -> return False
  - in each step update the current to the available TrieNode for the current character
  - at the end return .isEnd (for search, may be True or False) and return True for startsWith (prefix)
"""

class TrieNode:
    
    def __init__(self):
        
        self.children = [None] * 26
        self.isEnd = False

class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        currNode = self.root
        for i, char in enumerate(word):
            if currNode.children[ord(char) - ord("a")] == None:
                currNode.children[ord(char) - ord("a")] = TrieNode()
            currNode = currNode.children[ord(char) - ord("a")]
        
        currNode.isEnd = True
    
    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        currNode = self.root
        for i, char in enumerate(word):
            if currNode.children[ord(char) - ord("a")] == None: return False
            currNode = currNode.children[ord(char) - ord("a")]
            
        return currNode.isEnd

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        currNode = self.root
        for i, char in enumerate(prefix):
            if currNode.children[ord(char) - ord("a")] == None: return False
            currNode = currNode.children[ord(char) - ord("a")]
            
        return True  
        
# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
